argon2.go

```diff
@@ -0,0 +1,161 @@
+	// memory cost.
+	MemoryKiB int
+
+	// TargetDuration specifies the target duration for the KDF which
+	// is used to benchmark the time and memory cost parameters. If it
+	// is zero then the default is used. If ForceIterations is not zero
+	// then this field is ignored.
+	TargetDuration time.Duration
+
+	// ForceIterations can be used to turn off KDF benchmarking by
+	// setting the time cost directly. If this is zero then the cost
+	// parameters are benchmarked based on the value of TargetDuration.
+	ForceIterations int
+
+	// Parallel sets the maximum number of parallel threads for the
+	// KDF (up to 4). This will be adjusted downwards based on the
+	// actual number of CPUs.
+	Parallel int
+}
+
+func (o *KDFOptions) luksOpts() luks2.KDFOptions {
+	return luks2.KDFOptions{
+		TargetDuration:  o.TargetDuration,
+		MemoryKiB:       o.MemoryKiB,
+		ForceIterations: o.ForceIterations,
+		Parallel:        o.Parallel}
+}
+
+func (o *KDFOptions) deriveCostParams(keyLen int, kdf KDF) (*CostParams, error) {
+	switch {
+	case int64(o.ForceIterations) > math.MaxUint32:
+		return nil, errors.New("ForceIterations too large")
+	case int64(o.MemoryKiB) > math.MaxUint32:
+		return nil, errors.New("MemoryKiB too large")
+	case o.Parallel > math.MaxUint8:
+		return nil, errors.New("Parallel too large")
+	case o.ForceIterations < 0:
+		return nil, errors.New("ForceIterations can't be negative")
+	case int64(keyLen) > math.MaxUint32:
+		return nil, errors.New("keyLen too large")
+	case o.ForceIterations > 0:
+		threads := runtimeNumCPU()
+		if threads > 4 {
+			threads = 4
+		}
+		params := &CostParams{
+			Time:      uint32(o.ForceIterations),
+			MemoryKiB: 1 * 1024 * 1024,
+			Threads:   uint8(threads)}
+
+		if o.MemoryKiB != 0 {
+			params.MemoryKiB = uint32(o.MemoryKiB)
+		}
+		if o.Parallel != 0 {
+			params.Threads = uint8(o.Parallel)
+			if o.Parallel > 4 {
+				params.Threads = 4
+			}
+		}
+
+		return params, nil
+	default:
+		benchmarkParams := &argon2.BenchmarkParams{
+			MaxMemoryCostKiB: 1 * 1024 * 1024,
+			TargetDuration:   2 * time.Second}
+
+		if o.MemoryKiB != 0 {
+			benchmarkParams.MaxMemoryCostKiB = uint32(o.MemoryKiB)
+		}
+		if o.TargetDuration != 0 {
+			benchmarkParams.TargetDuration = o.TargetDuration
+		}
+		if o.Parallel != 0 {
+			benchmarkParams.Threads = uint8(o.Parallel)
+			if o.Parallel > 4 {
+				benchmarkParams.Threads = 4
+			}
+		}
+
+		params, err := argon2.Benchmark(benchmarkParams, func(params *CostParams) (time.Duration, error) {
+			return kdf.Time(params, uint32(keyLen))
+		})
+		if err != nil {
+			return nil, xerrors.Errorf("cannot benchmark KDF: %w", err)
+		}
+
+		return params, nil
+	}
+}
+
+// CostParams defines the cost parameters for key derivation using Argon2.
+type CostParams = argon2.CostParams
+
+// KDF is an interface to abstract use of the Argon2 KDF to make it possible
+// to delegate execution to a short-lived utility process where required.
+type KDF interface {
```

Comment:
shouldn't this be called Argon2KDF from the description? otherwise the fact that CostParams is actually argon2.CostParams is odd anyway

Reply:
We already have KDFOptions, so I'm not sure that renaming this to Argon2KDF without renaming KDFOptions makes sense.

What I've done instead is rename CostParams to KDFCostParams (and it's no longer type-aliased either), so we have KDF, KDFOptions and KDFCostParams.